edge/pkg/edgehub/upgrade/upgrade.go

```diff
@@ -0,0 +1,158 @@
+	}
+
+	return upgradeToolProviders[tool].Upgrade(upgradeReq)
+}
+
+func validateUpgrade(upgrade *commontypes.NodeUpgradeJobRequest) error {
+	if upgrade.UpgradeID == "" {
+		return fmt.Errorf("upgradeID cannot be empty")
+	}
+	if upgrade.Version == version.Get().String() {
+		return fmt.Errorf("edge node already on specific version, no need to upgrade")
+	}
+
+	return nil
+}
+
+type Provider interface {
+	Upgrade(upgrade *commontypes.NodeUpgradeJobRequest) error
+}
+
+var (
+	upgradeToolProviders = make(map[string]Provider)
+	mutex                sync.Mutex
+)
+
+func RegisterUpgradeProvider(upgradeToolType string, provider Provider) {
+	mutex.Lock()
+	defer mutex.Unlock()
+	upgradeToolProviders[upgradeToolType] = provider
+}
+
+type keadmUpgrade struct{}
+
+func (*keadmUpgrade) Upgrade(upgradeReq *commontypes.NodeUpgradeJobRequest) error {
+	// get edgecore start options and config
+	opts := options.GetEdgeCoreOptions()
+	config := options.GetEdgeCoreConfig()
+
+	// install the requested installer keadm from docker image
+	klog.Infof("Begin to download version %s keadm", upgradeReq.Version)
+	container, err := util.NewContainerRuntime(config.Modules.Edged.RuntimeType, config.Modules.Edged.RemoteRuntimeEndpoint)
+	if err != nil {
+		return fmt.Errorf("failed to new container runtime: %v", err)
+	}
+
+	imageRepository := upgradeReq.ImageRepository
+	// if users don't specify ImageRepository, we'll use default repository kubeedge
+	if imageRepository == "" {
+		imageRepository = "kubeedge"
+	}
+	image := image.EdgeSet(imageRepository, upgradeReq.Version).Get(image.EdgeCore)
+	// TODO: do some verification 1.sha256(pass in using CRD) 2.image signature verification
+	// TODO: release verification mechanism
+	err = container.PullImages([]string{image})
+	if err != nil {
+		return fmt.Errorf("pull image failed: %v", err)
+	}
+	files := map[string]string{
+		filepath.Join(util.KubeEdgeUsrBinPath, util.KeadmBinaryName): filepath.Join(util.KubeEdgeUsrBinPath, util.KeadmBinaryName),
+	}
+	err = container.CopyResources(image, files)
+	if err != nil {
+		return fmt.Errorf("failed to cp file from image to host: %v", err)
+	}
+
+	klog.Infof("Begin to run upgrade command")
+	upgradeCmd := fmt.Sprintf("keadm upgrade --upgradeID %s --historyID %s --fromVersion %s --toVersion %s --config %s --repository %s > /tmp/keadm.log 2>&1",
+		upgradeReq.UpgradeID, upgradeReq.HistoryID, version.Get(), upgradeReq.Version, opts.ConfigFile, imageRepository)
+
+	// run upgrade cmd to upgrade edge node
+	// use setsid command and nohup command to start a separate progress
+	command := fmt.Sprintf("setsid nohup %s &", upgradeCmd)
+	cmd := exec.Command("bash", "-c", command)
+	s, err := cmd.CombinedOutput()
+	if err != nil {
+		klog.Errorf("run upgrade command %s failed: %v, %s", command, err, s)
+		return fmt.Errorf("run upgrade command %s failed: %v, %s", command, err, s)
+	}
+
+	klog.Infof("!!! Begin to upgrade...")
```

Comment:
```suggestion
	klog.Infof("!!! Begin to upgrade from Version %s to %s ...", version.Get(), upgradeReq.Version)
```

Reply:
all above reviews fixed, except for those with objection reply, please check and review again